server/lib/api/users.js

```diff
@@ -96,3 +139,71 @@ async function resetPassword(ctx, next) {
     ctx.status = 204
   })
 }
+
+async function verifyEmail(ctx, next, userSockets) {
+  const { id } = ctx.params
+  const { code } = ctx.query
+  const { email } = ctx.request.body
+
+  if (!id || !code || !isValidEmail(email)) {
+    throw new httpErrors.BadRequest('Invalid parameters')
+  }
+
+  const emailVerified = await useEmailVerificationCode(id, email, code)
+  if (!emailVerified) {
+    throw new httpErrors.BadRequest('Email verification code is invalid')
+  }
+
+  ctx.status = 204
+
+  // If the user is not logged in, our job is done here. They will receive the updated data with
+  // `emailVerified` set to `true` next time they log in.
+  if (!ctx.session.userId) return
+
+  // If the user is logged in when verifying their email, update all of their sessions to indicate
+  // that their email is now indeed verified.
+  const userSessionsKey = 'user_sessions:' + ctx.session.userId
```

Comment:
You should probably pull all this session updating stuff out into a separate function so that this code doesn't proliferate all over the codebase. Something like `updateAllSessions(userId, updatedValues)` (the `sessionStore.set` call can then just pass `{ ...session, ...updatedValues }`)